File: sample-projects/kotlin/src/main/kotlin/de/sots/cellarsandcentaurs/adapter/persistence/Repository.kt
package de.sots.cellarsandcentaurs.adapter.persistence

abstract class Repository<T> {
    protected val store: MutableMap<String, T> = mutableMapOf()

    abstract fun keyOf(item: T): String

    fun save(item: T) {
        store[keyOf(item)] = item
    }

    fun findOne(key: String): T? = store[key]

    fun findAll(): List<T> = store.values.toList()
}

File: sample-projects/csharp/src/De.Sots.CellarsAndCentaurs/Adapter/Persistence/Repository.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace De.Sots.CellarsAndCentaurs.Adapter.Persistence;

public abstract class Repository<T> where T : class
{
    private readonly Dictionary<Guid, T> store = new();

    protected abstract Guid IdOf(T item);

    public void Add(T item) => store[IdOf(item)] = item;

    public T? FindOne(Guid id) => store.TryGetValue(id, out var item) ? item : null;

    public List<T> FindAll() => store.Values.ToList();
}
